src/coreclr/src/jit/earlyprop.cpp

```diff
@@ -477,208 +485,363 @@ GenTree* Compiler::optPropGetValueRec(unsigned lclNum, unsigned ssaNum, optPropK
+// optFoldNullChecks: Try to find a GT_NULLCHECK node that can be folded into the indirection node mark it for removal
+// if possible.
 //
 // Arguments:
-//    tree           - The input GT_INDIR tree.
+//    tree           - The input indirection tree.
+//    nullCheckMap   - Map of the local numbers to the latest NULLCHECKs on those locals in the current basic block
 //
+// Notes:
+//    If a GT_NULLCHECK node is post-dominated by an indirection node on the same local and the trees between
+//    the GT_NULLCHECK and the indirection don't have unsafe side effects, the GT_NULLCHECK can be removed.
+//    The indir will cause a NullReferenceException if and only if GT_NULLCHECK will cause the same
+//    NullReferenceException.
 
-void Compiler::optFoldNullCheck(GenTree* tree)
+void Compiler::optFoldNullCheck(GenTree* tree, LocalNumberToNullCheckTreeMap* nullCheckMap)
 {
-    //
-    // Check for a pattern like this:
-    //
-    //                         =
-    //                       /   \.
-    //                      x    comma
-    //                           /   \.
-    //                     nullcheck  +
-    //                         |     / \.
-    //                         y    y  const
-    //
-    //
-    //                    some trees in the same
-    //                    basic block with
-    //                    no unsafe side effects
-    //
-    //                           indir
-    //                             |
-    //                             x
-    //
-    // where the const is suitably small
-    // and transform it into
-    //
-    //                         =
-    //                       /   \.
-    //                      x     +
-    //                           / \.
-    //                          y  const
-    //
-    //
-    //              some trees with no unsafe side effects here
-    //
-    //                           indir
-    //                             |
-    //                             x
-
     if ((compCurBB->bbFlags & BBF_HAS_NULLCHECK) == 0)
     {
         return;
     }
 
-    assert(tree->OperIsIndir());
+    GenTree*   nullCheckTree   = optFindNullCheckToFold(tree, nullCheckMap);
+    GenTree*   nullCheckParent = nullptr;
+    Statement* nullCheckStmt   = nullptr;
+    if ((nullCheckTree != nullptr) && optIsNullCheckFoldingLegal(tree, nullCheckTree, &nullCheckParent, &nullCheckStmt))
+    {
+#ifdef DEBUG
+        if (verbose)
+        {
+            printf("optEarlyProp Marking a null check for removal\n");
+            gtDispTree(nullCheckTree);
+            printf("\n");
+        }
+#endif
+        // Remove the null check
+        nullCheckTree->gtFlags &= ~(GTF_EXCEPT | GTF_DONT_CSE);
+
+        // Set this flag to prevent reordering
+        nullCheckTree->gtFlags |= GTF_ORDER_SIDEEFF;
+        nullCheckTree->gtFlags |= GTF_IND_NONFAULTING;
+
+        if (nullCheckParent != nullptr)
```

Comment:
I pushed a change that added this `if` to deal with cases when the `GT_NULLCHECK` is the root node of the statement.